src/test/cpp/gtester/BatchRuns.cpp

```diff
@@ -193,7 +196,7 @@ TEST_P( BatchDemos, Run )
 	// Round up.
         // -----------------------------------------------------------------------------------------
 	const unsigned int num_cases = sim->GetPopulation()->get_infected_count();
-	ASSERT_NEAR(num_cases, g_results.at(test_tag),10000) << "!! CHANGED !!";
+	ASSERT_NEAR(num_cases, g_results.at(test_tag),15000) << "!! CHANGED !!";
```

Comment:
I thought about this for a while.

**TL;DR:** it's complicated and it's related to TBB's workload distribution policy.

## The root cause of the issue

There are three factors at play here.

  * OpenMP and our STL threads implementation are configured to statically schedule the workload. In other words, they carve up the vector to process and give each thread a slice. The `parallel_for` is done when all threads have processed their slices.

    TBB, on the other hand, distributes work across threads using a dynamic scheme, i.e., it gives each thread a chunk of elements to process, waits for them to complete and gives them more work.

    So if one thread gets slowed down by a resource-hungry program running on the same core as said thread, then TBB's default policy will make up for that by giving the other threads more elements to process. OpenMP and our STL thread implementation will just wait until the slow thread finishes. That said, I don't think this phenomenon amounts to a truly large performance difference in practice, but it is worth keeping in mind because it's one of TBB's selling points.

  * TBB doesn't have an equivalent of `omp_get_thread_num()`. This wouldn't have been problematic at all if Stride was designed with TBB in mind from the get-go, but currently Stride depends on these thread numbers to implement thread-local storage.

    For example, an `std::vector<RngHandler> m_rng_handler` of the same length as the number of threads in use is filled with random number generators at `Simulator` initialization-time and then indexed like so to get the random-number generator for the current thread: `rand[omp_get_thread_num()]`.

    I implemented thread numbers by dequeuing elements from a lock-based concurrent queue (`ConcurrentQueue<T>`). An interesting quirk of this implementation is that thread numbers may change: if two threads finish their workload simultaneously but one is restarted earlier than the other, then they will trade thread numbers.

    Note that this quirk *does not* break thread-local storage because no two threads will ever have the same thread number *at the same time.* It simply renames threads.

  * GitHub CI engines like Travis CI run Stride on a machine that's running other people's code, too. They sometimes even offer a fractional number of cores (usually 1.5). So we're basically getting one fast core and one slow core, which typically translate to one fast thread and one slow thread.

The interaction between these things is pretty interesting. Suppose that we have one fast thread and one sluggish thread. Because the former thread is so fast, it can finish later and still get restarted before the latter thread. The threads trade numbers as a consequence. They also trade thread-local storage. So *both* threads will now draw random numbers from *both* random-number generators and the total percentage of numbers drawn from the generators by each thread depends (among other things) on the difference in speed between the two threads.

Again, this *doesn't* break thread-local storage: to the application, the renumbering scheme makes it look like the threads are speeding up and slowing down all the time (they are actually getting renumbered). Each random-number generator will always remain tied to a thread number and, from the application's perspective, a thread *is* a thread number.

But TBB's workload distribution policy (and, probably, my thread renumbering scheme, too) does inject *nondeterminism* into the application: the random numbers assigned to each person vary based on the difference in data processing speed of TBB threads.

## The quick and easy fix and why it's a bad idea

We could easily change the workload balancing scheme used by TBB in the `parallel_for` to a static scheme, which will make it behave like our STL threads implementation. But that's a stupid solution because it reduces TBB to a glorified wrapper around system threads, *just like our STL threads implementation.* 

The whole point of TBB is that it's smart about how it distributes work. Using a static partitioning scheme destroys that smartness. What's the point of using TBB if we're going to cripple it so drastically that it won't even have a leg up on my 30 LOC STL threads `parallel_for` implementation?

## The smarter fix and why it's disingenuous

The `BatchRuns` test suite contains a call to `omp_set_schedule(omp_sched_static, 1);`. We could create a library-independent wrapper for `omp_set_schedule` and then set up a static workload distribution policy in `BatchRuns.cpp`. That'd probably help a great deal with determinism.

It'd also be a lie. Tests are a form of documentation. We'd be guaranteeing, in the test suite, that simulations with a setup similar to the `default` scenario in `BatchRuns` have an error margin of about `10000`. But even the `default` scenario itself doesn't have that error margin when dynamic scheduling is enabled!

So, in conclusion: an error margin of `15000` is a promise the simulator can probably keep. An error margin of `10000` is a promise the simulator can only keep in the controlled environment of a test suite with an artificially-constructed scheduling policy that is neither the default nor the fastest policy.

Also, the expected outcome of the `default` scenario was a bit of a lowball, so I've updated that.

P.S.: sorry about the long post.